pilot/model/controller_view.go

```diff
@@ -0,0 +1,580 @@
+	// by Port. While ServiceDiscovery function parms may contain
+	// alternate numeric formats for port numbers, ex: 80 or 0080
+	// etc, the value associated with this internal label is a
+	// normalized 16 bit int encoded in hex
+	labelInstancePort = labelInstancePrefix + "port"
+
+	// Label used internally by ControllerView to fetch instances
+	// by the named port.
+	labelInstanceNamedPort = labelInstancePrefix + "namedPort"
+)
+
+// ControllerView implements common storage of mesh entities like Service
+// and ServiceInstance objects and is intended to be used by individual
+// platform registry Controllers. Controller implementations call Reconcile()
+// to update the Controller's view with the most up-to-date list of mesh
+// entities available natively for that platform registry.
+//
+// TODO adopt ControllerView into all platform registry implementations as part
+// of https://github.com/istio/istio/issues/1223.
+//
+// Although objects from each Controller are distinct, they implicitly follow the
+// following contraints:
+//
+// - Service objects: are unique across the mesh for the combination of their
+// Hostname and Address
+// - ServiceInstance objects: are unique across the mesh for the combination
+// of their Service.Hostname and Endpoint.Address and Endpoint.Port
+//
+// The ControllerView will never modify model.* objects. However to accommodate
+// multiple data representations of various properties of model.* objects, the
+// ControllerView normalizes the following properties to allow for uniform
+// query operations across the mesh:
+//
+// - Address: Fetches by Service and ServiceInstance addresses are
+// always normalized to their 16 byte IPv6 representation.
+// - Port: Given that 0080 and 80 are equivalent, port numbers are always normalized
+// by their numeric 16 bit value.
+type ControllerView struct {
+
+	// serviceLabels holds the inverted map of label name/values to the keys in
+	// services for O(N) access involving labels versus O(N^2) access via services
+	serviceLabels viewLabels
+
+	// services is a map maintained by the Controller with a composite key of
+	// Service.Hostname and Service.Address. The key is expected to be mesh unique
+	services map[string]*Service
+
+	// instanceLabels holds the inverted map of label name/values to the keys in
+	// serviceInstances for O(N) access involving labels versus O(N^2) access via
+	// services
+	instanceLabels viewLabels
+
+	// serviceInstances is a map maintained by the Controller with a composite key
+	// of Service.Hostname, Endpoint.Address and Endpoint.Port. The key is expected
+	// to be mesh unique
+	serviceInstances map[string]*ServiceInstance
+
+	// Mutex guards updates to all model.* objects of the ControllerView
+	mu sync.RWMutex
+}
+
+// viewKeySet is a set of controller view keys. All keys in the
+// key set must be of the same type:
+//
+// - Service keys are of the form [Service.Hostname][16 byte hex formatted hex formatted
```

Comment:
Why ? Do we need both for key ? If it's a string, why create 32 byte in a representation nobody really uses ? Use cases for using both hostname and ip for lookup ?

Reply:
I don't think it matters. All these types and representation are entirely encapsulated.

Reply:
Of course it matters - encapsulation is not a reason to do arbitrary things. 
Go strings support bytes - so using the ip and port as 18 bytes would at least save one conversion, and take 1/2 the space. Still not explained why we even need to represent IP4 addresses in the 16-byte form in the key - it's not like having the same length matters. 

Data structures - and the keys - are one of the most important things in any app, in particular
if we want to scale up.

Reply:
I would like to avoid premature optimization whenever possible. But we need something to get started, so we can measure performance. One thing at a time :)